### PR TITLE
Fix run order of `didInsertParent` for children of already-set-up parents

### LINKED ISSUE #10: When a parent is already set up, `didInsertParent` fires before `didInsertElement`
This looks to be related to [parent.js:74](https://github.com/miguelcobain/ember-composability-tools/blob/master/addon/mixins/parent.js#L74) + the changes in eadbb005. 

Since the move to register children on `init()` instead of on `didInsertElement()`, children of an already-setup parent are receiving `didInsertParent()` on `init()`. This breaks any code in `didInsertParent()` that relies on the component's rendered element being present.

### PR DESCRIPTION
**Pull request by reidab**
Fixes #10

### CODE REVIEW

File: tests/integration/components/did-insert-parent-hook-test.js [hunk 1]
```diff
@@ -68,18 +77,19 @@ test('top-level parent and two children-parents', function(assert) {
   assert.ok(parentSpy.calledBefore(childSpy), 'parent was called before child');
 });
 
-test('top-level parent and two children-parents after if', function(assert) {
-  let parentSpy = this.parentSpy = this.spy();
-  let childSpy = this.childSpy = this.spy();
-  let childParentSpy = this.childParentSpy = this.spy();
+test('top-level parent with children-parents after if', function(assert) {
```
Comment (miguelcobain) [inline tests/integration/components/did-insert-parent-hook-test.js]: Why did you change existing tests rather than adding a new one?

File: addon/mixins/child.js [hunk 2]
```diff
@@ -55,11 +55,18 @@ export default Mixin.create({
     }
   },
 
+  notifyParentOfInsert: on('didInsertElement', function() {
```
Comment (miguelcobain) [inline addon/mixins/child.js]: We should use `didInsertParent` instead of `on('didInsertParent',`. It makes the calling order more predictable and provides a better opportunity for subclasses to override the method.

File: tests/integration/components/did-insert-parent-hook-test.js (hunk 1 above)
Comment (reidab) [inline tests/integration/components/did-insert-parent-hook-test.js]: I actually started out with it in a separate test. I moved it here because it wasn't testing a new thing, just clarifying of how the run order should work for parents that were already set up.

I'm happy to split it back out again if you prefer.

File: addon/mixins/child.js (hunk 2 above)
Comment (reidab) [inline addon/mixins/child.js]: I went with `on('didInsertElement'…` explicitly to avoid subclasses being able to override this and break the chain. I'll swap it.

File: tests/integration/components/did-insert-parent-hook-test.js (hunk 1 above)
Comment (reidab) [inline tests/integration/components/did-insert-parent-hook-test.js]: @miguelcobain Ping. Would you like me to split this out into a separate test?

Comment (miguelcobain) [inline tests/integration/components/did-insert-parent-hook-test.js]: Sorry for the delayed response, @reidab. Yes please, that would help a lot for me to understand what is going on here!

**Comment by reidab** [comment]
I'd completely forgotten about this PR until I went to update another library I had that depended on it. Apologies!

I've updated the PR to split out the new tests into their own test case and reverted back the changes to existing tests.

**Comment by reidab** [comment]
Ping. I've been delaying a library depending on this fix for many months now. Is there anything else I can do to move it forward?

**Comment by reidab** [comment]
Just rebased the branch on the current master resolved conflicts.

**Comment by Duder-onomy** [comment]
@miguelcobain I just made PR #19 that will fix the failures seen in this PR.

**Comment by reidab** [comment]
The description in @Duder-onomy's PR #19 said that it fixed this, so this got auto-closed, but it hasn't actually been merged. 😢

**Comment by Duder-onomy** [comment]
@miguelcobain This is still an issue, My PR accidentally closed this.. All my PR did was fix TravisCi. 
Can you re-open this? Or even better, merge it?

**Comment by Duder-onomy** [comment]
@miguelcobain 	Can this be merged?

**Comment by Duder-onomy** [comment]
@miguelcobain This is still an issue, i'm going to try and explain the issue the best I can. 

When the change was made to register children on `init()` instead of `didInsertElement()`, children of an already-setup parent ( the children were hidden behind some conditional ) are receiving `didInsertParent` on init, which breaks and code in the `didInsertParent` hook that relies on the element being available.

**Comment by Duder-onomy** [comment]
I believe this is the same issue seen in #16 and pull requested in #17.

**Comment by knownasilya** [comment]
Ping, just for visibility.

**Comment by miguelcobain** [comment]
The new architecture of ember-composability-tools is a lot different.
I added a modified version of this test to the current test suite and it passed. I can't keep the test though, because it required some hacking of the `Node` component to make the test possible (Good ol' times when we could just `willRender=this.spy` and it would end up replacing the component's willRender 😅 ).

In any case, I believe this is fixed, according to my tests. Feel free to reopen if this issue is found.
